Updating example to use stripe provided way of loading Stripe

mellson opened the pull request:
Hi @dayhaysoos I've tried to update the example to match the Stripe provided way of loading Stripe. I've updated and pinned some npm packages as well.

mellson commented:
Hi @dayhaysoos just added another commit to this which builds on yesterdays default values.
Based on the great input from @ChrisBrownie55 I think it makes sense to set `cartItems` to a sensible default value to get rid of the additional logic

mellson commented:
This doesn't change the behaviour in any way, so it doesn't require a release.

dayhaysoos commented:
@mellson 
This seems to break the dev environment in the example directory. I'm working on fixing it right now, just a heads up. For some reason Stripe was undefined in the netlify function, and now redirectToCheckout seems to be null here :eye:

ChrisBrownie55 commented on src/index.js:
```diff
@@ -112,7 +112,7 @@ function cartItemsReducer(cartItems, action) {
 }
 
-export const CartContext = createContext([{}, () => {}]);
+export const CartContext = createContext([{ cartItems: [] }, () => {}]);
```
I'd assume that if we're making sure there's a `cartItems` when there's no `CartProvider` that we should make sure everything else has its default value. You may want to define a variable above that has the default values for the rest of the cart (stuff handled by `cartReducer`) to avoid duplicating a bunch of lines in the code.

```suggestion
export const CartContext = createContext([{
  lastClicked: '',
  shouldDisplayCart: false,
  stripe,
  successUrl,
  cancelUrl,
  currency,
  billingAddressCollection,
  allowedCountries,
  skus: {},
  cartItems: [],
}, () => {}]);
```

dayhaysoos replied:
Oh! How obvious, can't believe I didn't think of that lol. Gonna test this out now

dayhaysoos commented:
@ChrisBrownie55 

Still having issues accessing redirectToCheckout, when I log the function before handleClick on the example, I can see the function there, no problem.

However, when handle click tries to use redirectToCheckout, I get:

`Unhandled Rejection (TypeError): Cannot read property 'redirectToCheckout' of null`

This is really bizarre. Any thoughts on this?

Also, I had to make these changes since they were default values and needed something to refer to. Perhaps I have to access the Stripe object without the tokenizing as default?

```js
export const CartContext = createContext([
  {
    cartItems: [],
    lastclicked: '',
    shouldDisplayCart: false,
    stripe: () => {},
    successUrl: '',
    cancelUrl: '',
    currency: '',
    billingAddressCollection: false,
    allowedCountries: [],
    skus: {},
    cartItems: [],
  },
  () => {},
]);
```

mellson commented:
@dayhaysoos I think the Netlify function had problems because I removed Stripe from the dependencies. Adding that again now.

dayhaysoos commented:
> @dayhaysoos I think the Netlify function had problems because I removed Stripe from the dependencies. Adding that again now.

NO WAIT DON'T DO IT I ALREADY DID

I want to figure out why the stripe object is remaining null within the CartProvider

mellson commented:
@dayhaysoos sorry, didn't see that before I did the commit. Can you try to see if it works now? I've changed the defaults to a minimal version where Gatsby can still build. If stripe or some of the other stuff that was undefined was in there the build broke

mellson commented:
@dayhaysoos I guess it's okay since they are not used at build time. They are only required when you actually have to use hook at runtime (at least I think so)

dayhaysoos commented:
@mellson when I run `yarn dev`, everything seems okay but once I try to checkout, redirectToCheckout is null. It seems like when the provider is passed the stripe object, the reducer never updates with the Stripe object. Can't seem to figure out why, still working on it

mellson commented:
Only a guess, but could it be the use of `useMemo` ? on line 165 in index.js ?

mellson commented:
I'll try to setup the Netlify cli tomorrow so I can test the full stack in the example folder

dayhaysoos commented:
@mellson it's actually a dev dependency, so not much will be needed for set up I hope. Would be cool to see if it works without much friction if you tried it directly from the package (if you don't mind experimenting)

mellson commented:
@dayhaysoos thanks, I've got it running now. And I have one working solution, but I want to test if there is a better one around the corner before committing

dayhaysoos commented:
@mellson interested in seeing what you got. I'm lost! lol

mellson commented:
@dayhaysoos so what happened was that using the Elements from Stripe to resolve the Stripe Promise caused CartProvider to be rendered twice. Once when stripe was null and once again when it wasn't null anymore. However CartProvider didn't update the state to match that stripe was no longer null.

I've tried to handle that in the last commit

dayhaysoos commented:
Ah, I see..makes sense, thank you. gonna test it out now

ChrisBrownie55 commented:
@mellson it could handle not having stripe by having a useEffect that, when the `stripe` prop changes a dispatch is fired to the `cartReducer` causing it to update the `stripe` object

ChrisBrownie55 commented:
Something like this should do:

```js
useEffect(() => {
  cartDispatch({ type: 'update stripe', stripe })
}, [stripe])
```

dayhaysoos commented:
This works!! 

I like @ChrisBrownie55 proposed refactor, using `useEffect`. Do you mind updating with that?

On a side note, we should also provide some testing for these instances with default values and null objects and what not. 

You've been very helpful @mellson, if you don't have time to write theses tests, I wouldn't mind doing it for you (but I'd open up an issue for later)

mellson commented:
I have updated the stripe handling to useEffect. I wouldn't mind doing the tests, but for some reason I can't run the tests in the project. I just get `Test suite failed to run. Cannot find module 'react' from 'state.js'`. I'll try to see if I can get them to run.

mellson commented:
I have added a few test cases now. I had to add React to the devDependencies of the root project for me to run the tests.

I also changed the useEffect for changing stripe slightly, so that it's not called too eagerly.

dayhaysoos commented:
@mellson oh wow! You actually fixed something I was hoping to create an issue for (regarding the react state errors with testing)

Going to pull this down and take a look

dayhaysoos commented:
just fyi I wasn't able to take a look yet, do plan on doing at some point today

dayhaysoos commented:
Gonna merge after @ChrisBrownie55 approval!

dayhaysoos commented:
@all-contributors please add @mellson for code

mellson commented:
Thanks for the kind help 👏👏👏

I’ve changed my mind about this not needing a release. Because if someone sees the updated example and tries to do that with v1.2.1 it won’t work as expected.

Have a great weekend when you get there!

ChrisBrownie55 commented:
A release would make sense yeah
